src/channel.cr

```diff
@@ -115,11 +118,17 @@ class Channel(T)
       raise_if_closed
 
       send_internal(value) do
-        @senders << Sender(T).new(Fiber.current, value, select_context: nil)
+        state = DeliveryState::None
+        @senders << Sender(T).new(Fiber.current, value, pointerof(state), select_context: nil)
         @lock.unsync do
           Crystal::Scheduler.reschedule
         end
-        raise_if_closed
+
+        if state == DeliveryState::Closed
+          raise ClosedError.new
+        elsif state != DeliveryState::Delivered
+          raise "BUG: Fiber was awaken without channel delivery state set"
+        end
```

Comment:
I'd rewrite it using `case`:

```cr
case state
when .closed?
  raise ClosedError.new
when .delivered?
  raise "BUG: Fiber was awaken without channel delivery state set"
end
```

Reply:
Nice suggestion. However, Channel#receive use `case` with the form
```crystal
case state
when DeliveryState::Delivered
  value
when DeliveryState::Closed
  yield
else
  raise "BUG: Fiber was awaken without channel delivery state set"
end
```

If use `case`, I think this will be more symmetric.
```crystal
case state
when DeliveryState::Delivered
  # ignore
when DeliveryState::Closed
  raise ClosedError.new
else
  raise "BUG: Fiber was awaken without channel delivery state set"
end
```

Reply:
Can you update the PR @firejox to the last snippet? `when DeliveryState::Closed`?

Reply:
@bcardiff sure